### PR TITLE
Runtime Cache Integration

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

related #681 

Changelog:

- add: `CachePreferenceAnalysis` that sets weight `1` to the right-hand-side expressions
- update: Enable caching in the `IdExecutionInstrument`
- update: Extend `InvalidateCache` command. All cache invalidation is performed with `InvalidateCache#run` methods before the program execution

### Important Notes

To complete the integration task, the `Changeset` algorithm should be updated to support batch edits.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/docs/style-guide/scala.md) and [Java](https://github.com/luna/enso/blob/master/docs/style-guide/java.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
    * @return {@code true} if the value was added to the cache.
    */
   public boolean offer(UUID key, Object value) {
-    if (weights.getOrDefault(key, 0L) > 0) {
+    if (weights.getOrDefault(key, 0.0) > 0) {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java:23]: This should depend on the constants from the weighting pass, not encode a literal.

File: engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/CachePreferenceAnalysis.scala [hunk 2]
```diff
@@ -21,17 +32,25 @@ case object CachePreferenceAnalysis extends IRPass {
 
   override type Metadata = WeightInfo
 
-  /** @inheritdoc **/
+  /** Run deshugaring passes first. */
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/CachePreferenceAnalysis.scala:35]:
```suggestion
  /** Run desugaring passes first. */
```

File: engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/CachePreferenceAnalysis.scala [hunk 3]
```diff
@@ -106,140 +131,25 @@ case object CachePreferenceAnalysis extends IRPass {
...
-        block
-          .copy(
-            expressions = expressions.map(analyseExpression(_, weights)),
-            returnValue = analyseExpression(returnValue, weights)
-          )
-          .updateMetadata(this -->> weights)
-
-      case binding @ IR.Expression.Binding(name, expression, _, _, _) =>
-        expression.getExternalId.foreach(weights.update(_, 1L))
+    expression.transformExpressions {
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/CachePreferenceAnalysis.scala:134]: This doesn't currently assign weights to non-bindings, even if they have external identifiers. It should.

Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/CachePreferenceAnalysis.scala:134]: Forgot about it. Fixed :+1:

File: engine/runtime/src/main/scala/org/enso/compiler/codegen/IRToTruffle.scala [hunk 4]
```diff
@@ -169,14 +170,19 @@ class IRToTruffle(
             DataflowAnalysis,
             "No dataflow information associated with an atom."
           )
+          val cacheInfo = atomDefn.unsafeGetMetadata(
```
Comment (kustosz) [inline engine/runtime/src/main/scala/org/enso/compiler/codegen/IRToTruffle.scala:173]: This is all very strange. What's the reason for storing these in the Local Scope in the first place? And, even more importantly, what is the point of storing them per Local Scope if this object is a module-level singleton? Something doesn't make sense here.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/BaseVisualisationCmd.scala [hunk 5]
```diff
@@ -65,6 +65,13 @@ abstract class BaseVisualisationCmd
           Api.Response(requestId, replyWith)
         )
         val stack = ctx.contextManager.getStack(config.executionContextId)
+        CacheInvalidation.run(
+          stack,
+          CacheInvalidation(
+            CacheInvalidation.StackSelector.Top,
+            CacheInvalidation.Command.InvalidateKeys(Seq(expressionId))
```
Comment (kustosz) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/BaseVisualisationCmd.scala:72]: This is wrong. If the value is cached, don't remove it, just reuse for visualization.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/PopContextCmd.scala [hunk 6]
```diff
@@ -1,13 +1,19 @@
...
-import org.enso.interpreter.instrument.InstrumentFrame
+import org.enso.interpreter.instrument.{CacheInvalidation, InstrumentFrame}
 import org.enso.interpreter.instrument.execution.RuntimeContext
 import org.enso.polyglot.runtime.Runtime.Api
 import org.enso.polyglot.runtime.Runtime.Api.RequestId
 
 /**
   * A command that pops an item from a stack.
   *
+  * == Caching ==
```
Comment (kustosz) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/PopContextCmd.scala:11]: I don't understand this logic. The popped frame contains values specific to the popped frame, right? So what's the point retaining them?

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/BaseVisualisationCmd.scala (hunk 5 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/BaseVisualisationCmd.scala:72]: Legit. I added the ability to update visualizations without recomputing the expression :+1:

File: engine/runtime/src/main/scala/org/enso/compiler/codegen/IRToTruffle.scala (hunk 4 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/compiler/codegen/IRToTruffle.scala:173]: Ok, I was not sure how `CachePreferenceAnalysys#runExpression` should be implemented. Turned out we don't need the metadata in the local scope. Removed.

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java [hunk 7]
```diff
@@ -162,17 +132,25 @@ public void onEnter(EventContext context, VirtualFrame frame) {
...
         nodeId = ((ExpressionNode) node).getId();
+      } else if (node instanceof FunctionCallInstrumentationNode) {
+        nodeId = ((FunctionCallInstrumentationNode) node).getId();
       }
 
-      Object overrideValue = cache.get(nodeId);
-      if (overrideValue != null) {
-        throw context.createUnwind(overrideValue);
+      Object result = cache.get(nodeId);
+      if (result != null) {
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:146]: Won't this recompute visualizations for all cached values? That sounds wrong. As we spoke before, this instrument should have two different modes of operation:
1. Compute a specific set of IDs – getting it from cache if exists, skipping unnecessary computations if possible etc.
2. Recompute everything that's changed (that's what we already have).
I'm not sure how to implement it in the most elegant way, but that's what we need from the feature.

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java [hunk 8]
```diff
@@ -190,19 +168,17 @@ public void onReturnValue(EventContext context, VirtualFrame frame, Object resul
...
       if (node instanceof FunctionCallInstrumentationNode
           && result instanceof FunctionCallInstrumentationNode.FunctionCall) {
-        ExpressionCall expressionCall =
-            new ExpressionCall(
-                ((FunctionCallInstrumentationNode) node).getId(),
-                (FunctionCallInstrumentationNode.FunctionCall) result);
-        calls.put(expressionCall.getExpressionId(), expressionCall.getCall());
-        functionCallCallback.accept(expressionCall);
+        UUID nodeId = ((FunctionCallInstrumentationNode) node).getId();
+        cache.putEnterable(nodeId, (FunctionCallInstrumentationNode.FunctionCall) result);
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:175]: I'm not sure we can safely cache enterables. We definitely should cache the currently-entered-enterable, but keeping all of them is dangerous – they contain function arguments and these can be arbitrarily big.

Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:175]: So this is probably best modeled by a per-cache-instance singleton? Something like "this cache is generated by entering such and such function".

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java (hunk 7 above)
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:146]: Oh I've just read some more and it seems like the implementation actually does what I expected :P.
It is a bit confusing, but it works properly, so disregard the above.

**Comment by kustosz** [comment]
> Awesome stuff in general. Only problem is, I don't think we can just cache the enterables like this – it's unbound memory again.

Though we definitely must cache the currently-entered enterable so there's some thinking to do :)

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java (hunk 8 above)
Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java]: OK, fixed, now we cache only the top stack frame, the only one we care about.

**Pull request merged**